src/irmin/tree.ml

```diff
@@ -668,6 +668,17 @@ module Make (P : Private.S) = struct
             let entries = P.Node.Val.list v in
             List.for_all (fun (step, _) -> StepMap.mem step um) entries)
 
+    let length t =
+      match cached_map t with
+      | Some m -> StepMap.cardinal m |> Lwt.return
+      | None -> (
+          let+ v = to_value t in
+          match v with
+          | Ok v -> P.Node.Val.length v
+          | Error (`Dangling_hash hash) ->
+              Fmt.failwith "length: encountered dangling hash %a"
+                (Type.pp P.Hash.t) hash)
```

Comment:
This is almost https://github.com/mirage/irmin/blob/b0e62ef7afd766e3b1324032636219524265ff25/src/irmin/tree.ml#L126-L129, which you can extend with a function context if you want it.

Reply:
I'll push the `get_ok` with context in a second PR